### PR TITLE
Fix dev container

### PR DESCRIPTION
**Pull request by amirmolavi**
Now this should work and it feels right. 

From the beginning all of the followings felt off but I just couldn't see the right solution:
1) we copied everything in `/` and the we mount local we the same files.
2) after committing there were two environments one in `/` and another in `/workspaces/llamabot/`
3) you could not commit to the repo from your local and within the container. pre commit would fail.

All of these issues meant that `.pixi` should live in the mounted folder but I just couldn't figure it out. I knew that .pixi shouldn't be on a case insensitive filesystem but instead on its own volume. so we needed to map it. (https://pixi.sh/latest/ide_integration/devcontainer/)

### CODE REVIEW

**Comment by amirmolavi** [comment]
it never gets easier :D 
why did you need to pin down the version of `sqlite-vec`? it does not support linux-aarch64

```
  x failed to solve the pypi requirements of 'notebooks' 'linux-aarch64'
  |-> failed to resolve pypi dependencies
  `-> Because sqlite-vec==0.1.3 has no wheels with a matching platform tag
      and only sqlite-vec<=0.1.3 is available, we can conclude that sqlite-
      vec>=0.1.3 cannot be used.
      And because you require sqlite-vec>=0.1.3, we can conclude that your
      requirements are unsatisfiable.
      
      hint: Pre-releases are available for sqlite-vec in the requested
      range (e.g., 0.1.4a2), but pre-releases weren't enabled (try:
      `--prerelease=allow`)
```

**Comment by amirmolavi** [comment]
> Aww shucks! I thought using pixi would make it easier haha.
> 
> Looks like I should revert my addition of sqlite-vec. It was an experiment, but the thing is very bleeding edge.
> 
> I wonder if there might be a way to make sqlite-vec work only on supported systems? Just thinking aloud. Perhaps the better idea is to just undo it, which I’m fine with. Once I put the kids to bed, I’ll do a new PR that makes sqlite-vec purely optional. That should do the trick, I hope. Once you have that merged in, we can see how this PR evolves.
> 
> Super appreciative of your patience here! We will get it over the finish line!

No don't, I am trying to set it up on an amd platform instead of arm. looks promising.

**Comment by amirmolavi** [comment]
Okay, everything (almost everything) is all set. I am really happy with this container setup. it is clean. Only the docker login fails in the workflow. I checked the owner and it is you, I am not sure why it is failing.

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM ghcr.io/prefix-dev/pixi:latest
+FROM --platform=linux/amd64 ghcr.io/prefix-dev/pixi:latest
```
Comment (ericmjl) [inline .devcontainer/Dockerfile:1]: Wow, this is a great one! Today I learned something new!

**Comment by ericmjl** [comment]
This is awesome, thank you @amirmolavi! I learned so much from how you set it up. May I ask for one final favour: could you add documentation on the design decisions you took w.r.t. this devcontainer? That will help us for the future. One example question is why you chose to move the pixi setup steps out of the container and into the `devcontainer.json` file.

We can ignore the "Build development container" action, it requires creds that only exist on my account, so the Action will fail when it's run from your account (which is what's happening right now).

Once we have the documentation pushed, we can merge! And I'd love for you to do the honors of hitting the big green button :smile:.

**Comment by amirmolavi** [comment]
> This is awesome, thank you @amirmolavi! I learned so much from how you set it up. May I ask for one final favour: could you add documentation on the design decisions you took w.r.t. this devcontainer? That will help us for the future. One example question is why you chose to move the pixi setup steps out of the container and into the `devcontainer.json` file.
> 
> We can ignore the "Build development container" action, it requires creds that only exist on my account, so the Action will fail when it's run from your account (which is what's happening right now).
> 
> Once we have the documentation pushed, we can merge! And I'd love for you to do the honors of hitting the big green button 😄.



> This is awesome, thank you @amirmolavi! I learned so much from how you set it up. May I ask for one final favour: could you add documentation on the design decisions you took w.r.t. this devcontainer? That will help us for the future. One example question is why you chose to move the pixi setup steps out of the container and into the `devcontainer.json` file.
> 
> We can ignore the "Build development container" action, it requires creds that only exist on my account, so the Action will fail when it's run from your account (which is what's happening right now).
> 
> Once we have the documentation pushed, we can merge! And I'd love for you to do the honors of hitting the big green button 😄.



> This is awesome, thank you @amirmolavi! I learned so much from how you set it up. May I ask for one final favour: could you add documentation on the design decisions you took w.r.t. this devcontainer? That will help us for the future. One example question is why you chose to move the pixi setup steps out of the container and into the `devcontainer.json` file.
> 
> We can ignore the "Build development container" action, it requires creds that only exist on my account, so the Action will fail when it's run from your account (which is what's happening right now).
> 
> Once we have the documentation pushed, we can merge! And I'd love for you to do the honors of hitting the big green button 😄.

Sure thing, I learned most of these in the past week. before this PR I only used containers and pixi just as a user and never set them up myself. 
For your particular example, I did not have any choice but to move the install step to post creation. The reason is the local repo is being mounted on /workspaces/llamabot after the creation. Dockerfile does not have access to that folder. I think the main decision and made everything that fall into place was that there should be one .pixi and it should be in the mounted folder. I will go through the documentations and make sure they are accurate and add details if needed.

**Comment by amirmolavi** [comment]
@ericmjl I merged my branch this morning but some tests didn't pass (loading the LLM looks flaky). Moreover, I found one issue.
1) The dev container is still broken. since it was broken before, please decide if you wanna revert my merge or do you wanna me find a fix on top of the master.
2) The problem is related to the `amd` platform. I am not sure how, but in the `amd` platform pixi mess up the numpy installation and running the tests in the dev environment fails. 
3) `arm` platform is working fine but then the problem with `sqlit-vec` is coming up.
4) I opened a slack channel and please feel free to ignore it if you don't want ant additional noise in your days. I thought it might be easier to do some communications there. I invited you via your @hey email but here is the link if you wanna join with another email address. (https://join.slack.com/t/llamabotworkspace/shared_invite/zt-2tk403zg2-tap5aLo1nEiB6ptyX52RWw)
5) I am still looking into these to find a solution to make this work and also keep sqlite-vec.
6) Sorry for the inconvenience and all the noise and emails.

**Comment by ericmjl** [comment]
@amirmolavi no worries! I appreciate the dedication here. 

I think we can archive the sqlite vec stuff. It’s a bit too early and bleeding edge, I think. It’s no hard feelings here too, as I was just playing around with it. I’d be happy to go back to a pyproject.toml file that omits sqlitevec to make it work.

**Comment by ericmjl** [comment]
As for what to do with the dev container, I think let’s try to debug it without sqlite-vec. Would you be open to that?

**Comment by amirmolavi** [comment]
@ericmjl I love that. I think that fixing the dev environment is a very critical step. however, It needs more than just fixing the toml file since there are tests and code for that. before you doing so, let me create a branch from the main before sqlite-vec and apply my changes and confirm its functionality.

**Pull request merged**